### PR TITLE
Fine tune unit tests to improve reliability

### PR DESCRIPTION
**Pull request by bwaresiak**
Task/Issue URL: https://app.asana.com/0/856498667320406/1208724390819804/f
Tech Design URL:
CC:

**Description**:

- Separate out WebView/WebKit tests to a new target.
- Add long-living WebView to host app view hierarchy.
- Disable flaky tests.
- Tweak how we wait for expectations when structured concurrency is involved.

**Steps to test this PR**:

Run tests :)

**Definition of Done (Internal Only)**:

* [ ] Does this PR satisfy our [Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f)?

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 15
* [ ] iOS 16
* [ ] iOS 17

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: DuckDuckGoTests/SyncUI/SyncManagementViewModelTests.swift [hunk 1]
```diff
@@ -159,7 +159,7 @@ class SyncManagementViewModelTests: XCTestCase, SyncManagementViewModelDelegate
     }
 
 
-    func testWhenRecoverSyncDataPressed_RecoverDataViewShown() {
+    func testWhenRecoverSyncDataPressed_RecoverDataViewShown() throws {
```
Comment (samsymons) [inline DuckDuckGoTests/SyncUI/SyncManagementViewModelTests.swift]: Is this change still needed?

File: WebViewUnitTests/UserAgentTests.swift [hunk 2]
```diff
@@ -23,6 +23,53 @@ import XCTest
...
+    var atb: String?
+    var searchRetentionAtb: String?
+    var appRetentionAtb: String?
+
+    var hasInstallStatistics: Bool {
+        return atb != nil
+    }
+
+    var variant: String?
+}
```
Comment (samsymons) [inline WebViewUnitTests/UserAgentTests.swift:38]: Some of these mocks are duplicated, we could reduce some of that by adding relevant test helpers to this new `WebViewUnitTests` target, but tbh I wouldn't block the PR on it - it's acceptable as-is.

File: WebViewUnitTests/UserDefaultsFireproofingTests.swift [hunk 3]
```diff
@@ -20,6 +20,44 @@
 import XCTest
 @testable import Core
 
+// TODO: More broad shared TestUtils
```
Comment (samsymons) [inline WebViewUnitTests/UserDefaultsFireproofingTests.swift:23]: Could be a good Special Days task for someone? :)

File: DuckDuckGoTests/FireButtonReferenceTests.swift [hunk 4]
```diff
@@ -63,11 +64,15 @@ final class FireButtonReferenceTests: XCTestCase {
...
 
+        let warmup = WebViewWarmupHelper()
+
         for test in referenceTests {
             let cookie = try XCTUnwrap(cookie(for: test))
 
-            let warmup = DataStoreWarmup()
-            await warmup.ensureReady(applicationState: .unknown)
+            let warmupCompleted = XCTestExpectation(description: "Warmup Completed")
+            warmup.warmupWebView(expectation: warmupCompleted)
```
Comment (samsymons) [inline DuckDuckGoTests/FireButtonReferenceTests.swift:72]: This will check `XCTAssertNil(webView)`, but since we're running this in the loop, it'll hit it on the second iteration since we're not resetting the webview anywhere (and I don't think we want to) - I can reproduce this locally every time, but CI seems happy with it 🤔

**Review by samsymons — APPROVED** [review]
LGTM, added one note about the `warmupWebView` logic since it causes a consistent failure for me locally (but not on CI?), otherwise just some small cosmetic comments. Nice work!

File: DuckDuckGoTests/FireButtonReferenceTests.swift (hunk 4 above)
Comment (bwaresiak) [reply DuckDuckGoTests/FireButtonReferenceTests.swift:72]: Good catch, it seems for some reason this test was disabled in https://github.com/duckduckgo/iOS/commit/f8fc6c5a486ee86da777447e23cf70ab4a31fd30

Comment (bwaresiak) [reply DuckDuckGoTests/FireButtonReferenceTests.swift:72]: Ok I've spoke with @brindy and as part of his work he is going to remove and tweak these tests so I'm going to ignore this problem for now.

File: WebViewUnitTests/UserDefaultsFireproofingTests.swift (hunk 3 above)
Comment (bwaresiak) [reply WebViewUnitTests/UserDefaultsFireproofingTests.swift:23]: Maybe? :)

File: WebViewUnitTests/UserAgentTests.swift (hunk 2 above)
Comment (bwaresiak) [reply WebViewUnitTests/UserAgentTests.swift:38]: I agree - hence my comment below.

File: DuckDuckGoTests/FireButtonReferenceTests.swift (hunk 4 above)
Comment (samsymons) [reply DuckDuckGoTests/FireButtonReferenceTests.swift:72]: Sounds good, thanks!

**Comment by samsymons** [comment]
New changes look good, thanks @bwaresiak! I re-ran the workflows a few times and everything was green every time, great work!
